rholang/src/main/k/rholang.k

```diff
@@ -1,132 +1,1039 @@
 module RHO-SYNTAX
 imports SUBSTITUTION
 
-// Syntax for processes
-syntax Proc ::=
-              // Empty process
-                "Nil"
+// We need processes, including variables
+syntax Proc ::= ProcNoVars
+              | Id
+
+// We need processes, excluding variables
+syntax ProcNoVars ::=
+              // Ground terms, including the empty process and expressions
+                Ground
               // Listen
-              | "for(" Name "<-" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<-" Name ")" "{" Proc "}" [binder]
               // Persistent listen
-              | "for(" Name "<=" Name "){" Proc "}" [binder]
+              | "for" "(" Names "<=" Name ")" "{" Proc "}" [binder]
               | "contract" Name "(" Name ")" "=" "{" Proc "}" [binder]
```

Comment:
Shouldn't this (and the line below) be `[binder(2 -> 3)]`? Otherwise in `contract x(y){ P }`, `x` is bound in `P` (and `y` it seems).

> By default, binder binds all the variables occurring anywhere in the first argument of the corresponding syntactic construct within its other arguments.

Reply:
You're right, good catch. Thank you. I'm going to write `"for" "(" Names "<=" Name ")" "{" Proc "}" [binder(1 -> 3)]` and then take the binder off of the `"contracts"` construct—later on in the framework there's a rule showing that `"contract"` is just syntactic sugar for the persistent send, and when it gets rewritten it will bind the variables appropriately.

Reply:
(I think it should be `[binder(1 -> 3)]`, not `[binder(2 -> 3)]` for the persistent listen.)

Reply:
`[binder(1 -> 3)]` for the `for` form, `[binder(2 -> 3)]` for the `contract` form.

Reply:
Also, now that I am looking at it, line 16 should contain `Names` instead of `Name` inside the parens, right?

Reply:
No, contract is syntactic sugar for the persistent `for` but when we only listen for one pattern (and not a ≥2-tuple of patterns).

Reply:
Hmm, I thought `contract @"hello"(firstName, lastName) { P }` and `for ( firstName, lastName <= @"hello" ) { P }` were both valid (and equivalent) constructions.

Reply:
They are.  He meant that there's no way of sugaring `for (i <=x; j <= y) { P }` into a contract.

Reply:
Ah, right, I wasn't quite clear on that; in that case the current framework is indeed incorrect

Reply:
```
"for" "(" Names "<=" Name ")" "{" Proc "}" [binder]
"contract" Name "(" Name ")" "=" "{" Proc "}" [binder]
```
I'm still seeing an error. I understand that the "join" syntax doesn't work in the `contract` form, but if `for ( a, b <= x ) { P }` is supported by these productions, then `contract x( a, b ) = { P }` should also be supported.

Reply:
Yes, we all agree that the second `Name` should be `Names` now and eventually `Patterns` later (or whatever the right production is).